src/util/psbt/map/global.rs

```diff
@@ -0,0 +1,133 @@
+    pub unknown: HashMap<raw::Key, Vec<u8>>,
+}
+
+impl Global {
+    /// Create a Global from an unsigned transaction, error if not unsigned
+    pub fn from_unsigned_tx(tx: Transaction) -> Result<Self, encode::Error> {
+        for txin in &tx.input {
+            if !txin.script_sig.is_empty() {
+                return Err(Error::UnsignedTxHasScriptSigs.into());
+            }
+
+            if !txin.witness.is_empty() {
+                return Err(Error::UnsignedTxHasScriptWitnesses.into());
+            }
+        }
+
+        Ok(Global {
+            unsigned_tx: tx,
+            unknown: Default::default(),
+        })
+    }
+}
+
+impl Map for Global {
+    fn insert_pair(&mut self, pair: raw::Pair) -> Result<(), encode::Error> {
+        let raw::Pair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(Error::DuplicateKey(raw_key).into());
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(Error::DuplicateKey(raw_key).into());
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<raw::Pair>, encode::Error> {
+        let mut rv: Vec<raw::Pair> = Default::default();
+
+        rv.push(raw::Pair {
+            key: raw::Key {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: self.unsigned_tx.serialize(),
+        });
+
+        for (key, value) in self.unknown.iter() {
+            rv.push(raw::Pair {
+                key: key.clone(),
+                value: value.clone(),
+            });
+        }
+
+        Ok(rv)
+    }
+
+    fn merge(&mut self, other: Self) -> Result<(), psbt::Error> {
+        if self.unsigned_tx != other.unsigned_tx {
+            return Err(psbt::Error::UnexpectedUnsignedTx {
+                expected: self.unsigned_tx.clone(),
+                actual: other.unsigned_tx,
+            });
+        }
+
+        self.unknown.extend(other.unknown);
+        Ok(())
+    }
+}
+
+impl_psbtmap_consensus_encoding!(Global);
+
+impl<D: Decoder> Decodable<D> for Global {
+    fn consensus_decode(d: &mut D) -> Result<Self, encode::Error> {
+        let mut pairs: Vec<raw::Pair> = Default::default();
+
+        while let Ok(pair) = Decodable::consensus_decode(d) {
+            pairs.push(pair);
+        }
+
+        let (tx_raw_kv_pair, unknown_raw_kv_pairs): (
+            Vec<raw::Pair>,
+            Vec<raw::Pair>,
+        ) = pairs
+            .into_iter()
+            .partition(|ref raw_kv_pair| raw_kv_pair.key.type_value == 0u8);
+
+        match tx_raw_kv_pair.len() {
+            0 => Err(Error::MustHaveUnsignedTx.into()),
+            1 => {
+                let tx: Transaction = deserialize(&tx_raw_kv_pair[0].value[..])?;
```

Comment:
Can you squash the "decode transaction manually" commit into this?

Also, you need to check that there are no trailing bytes.

Reply:
Done and checked: https://github.com/rust-bitcoin/rust-bitcoin/pull/103/commits/99da3e620043cf6d88e1b5067c33b48fdb6d9e37#diff-6e12ff92fec0d0252713934f51fe80fcR126